writer.go

```diff
@@ -1,37 +1,215 @@
+	// the returned value. Use this only if you don't care about guarantees of
+	// whether the messages were written to kafka.
+	Async bool
+
+	// An optional function called when the writer succeeds or fails the
+	// delivery of messages to a kafka partition. When writing the messages
+	// fails, the `err` parameter will be non-nil.
+	//
+	// The messages that the Completion function is called with have their
+	// topic, partition, offset, and time set based on the Produce responses
+	// received from kafka. All messages passed to a call to the function have
+	// been written to the same partition. The keys and values of messages are
+	// referencing the original byte slices carried by messages in the calls to
+	// WriteMessages.
+	//
+	// The function is called from goroutines started by the writer. Calls to
+	// Close will block on the Completion function calls. When the Writer is
+	// not writing asynchronously, the WriteMessages call will also block on
+	// Completion function, which is a useful guarantee if the byte slices
+	// for the message keys and values are intended to be reused after the
+	// WriteMessages call returned.
+	//
+	// If a completion function panics, the program terminates because the
+	// panic is not recovered by the writer and bubbles up to the top of the
+	// goroutine's call stack.
+	Completion func(messages []Message, err error)
+
+	// Compression set the compression codec to be used to compress messages.
+	Compression Compression
+
+	// If not nil, specifies a logger used to report internal changes within the
+	// writer.
+	Logger Logger
+
+	// ErrorLogger is the logger used to report errors. If nil, the writer falls
+	// back to using Logger instead.
+	ErrorLogger Logger
+
+	// A transport used to send messages to kafka clusters.
+	//
+	// If nil, DefaultTransport is used.
+	Transport RoundTripper
+
+	// Atomic flag indicating whether the writer has been closed.
+	closed uint32
+	group  sync.WaitGroup
+
+	// Manages the current batch being aggregated on the writer.
+	mutex   sync.Mutex
+	batches map[int32]*writeBatch
 
 	// writer stats are all made of atomic values, no need for synchronization.
-	// Use a pointer to ensure 64-bit alignment of the values.
-	stats *writerStats
+	// Use a pointer to ensure 64-bit alignment of the values. The once value is
+	// used to lazily create the value when first used, allowing programs to use
+	// the zero-value value of Writer.
+	once sync.Once
+	*writerStats
+
+	// If no balancer is configured, the writer uses this one. RoundRobin values
+	// are safe to use concurrently from multiple goroutines, there is no need
+	// for extra synchronization to access this field.
+	roundRobin RoundRobin
+
+	// non-nil when a transport was created by NewWriter, remove in 1.0.
+	transport *Transport
 }
 
 // WriterConfig is a configuration type used to create new instances of Writer.
+//
+// DEPRECATED: writer values should be configured directly by assigning their
```

Comment:
I'm questioning whether doing a two-phase deprecation is worthwhile, it seems OK to remove API while in v0. seems like we're delaying ripping off the band-aid :)

Reply:
but I'm fully open to being wrong about that of course

Reply:
I thought about it, but I feel like we're doing the right thing by optimizing for backward compatibility. I've had too many pains with library's making breaking changes, it's true that we are below 1.0 but it doesn't mean we shouldn't be mindful of the extra work we push down to users when we make breaking changes. There are a few breaking changes in 0.4 already (all trivial to address), I'd like to keep the count as low as possible and do the API cleanup when we're ready to release 1.0.